### DISCUSSION TITLE
Mass thumbnail generation - memory issues and crashes

### DISCUSSION THREAD

**Discussion by andy250** [libvips/libvips#3484] [Q&A]
Hello devs,
We are using libvips 8.14 (self-compiled) for a lot of thumbnail work on windows servers (2019 and 2022). We create every thumb by spawning new vips.exe process. On some environments it runs into memory issues. Sample messages from event viewer:

`Windows successfully diagnosed a low virtual memory condition. The following programs consumed the most virtual memory: vips.exe (12480) consumed 35432935424 bytes, (...)`

It's weird that single thumbnail run would eat up that much. Is vips doing some magic to try limit number of processes that are doing the main work? So spawning N processes puts all the work into some other vips.exe process?

Moments later there are some crashes reported (obviously, since we're out of memory):

```
Faulting application name: vips.exe, version: 0.0.0.0, time stamp: 0x63c8034f
Faulting module name: vips-magick.dll, version: 0.0.0.0, time stamp: 0x63c80345
Exception code: 0xc0000005
Fault offset: 0x000000000000276b
Faulting process id: 0x30c0
Faulting application start time: 0x01d981777cb6a02f
Faulting application path: C:\libvips\bin\vips.exe
Faulting module path: C:\libvips\bin\vips-modules-8.14\vips-magick.dll
Report Id: 9dd78a55-86cf-40a7-9dcc-ba51de6d6f8f
Faulting package full name:
Faulting package-relative application ID: 
```

This later causes more trouble for the entire server and other apps that are running there (some custom windows services get stuck and can't be killed etc.). My main question: could this be caused by a bug in libvips? Or is it just our use-case - spawning too much thumbnail requests leads to this? Images we work with are mostly jpg, pdf, png sometimes heic. Let me know if there is any more diagnostic info I could provide to nail this down.

EDIT: is there any way to tell vips to limit resources it is using? (threads/memory/priorities)

**Comment by jcupitt** [comment]
Hi @andy250,

Is the high memory use random, or is it triggered by specific images? If it's a specific image, can you send a sample?

You can get issues with memory bombs -- for example, it's easy to make a TIFF image which is only a few 100 kb, but which will need many GB to thumbnail. The best solution is usually to run `vipsheader` first and reject any images which have `width * height * bands > threshold`.

**Comment by jcupitt** [comment]
I made an example:

```
$ vips black x.tif[compression=ccittfax4,bitdepth=1] 10000000 10000
$ ls -l x.tif
-rw-rw-r-- 1 john john 213124 May  8 13:42 x.tif
$ /usr/bin/time -f %M:%e vipsthumbnail x.tif 
38112124:501.36
```

So a 200kb image takes 10 minutes of CPU and 4gb of memory to thumbnail.

It's also worth checking for things like progressive JPEG and PNG, and making sure you've removed dependencies like imagemagick.

**Reply by andy250** [reply]
Hmm, can you elaborate on removing the dependencies? I just complied it for windows according to instructions to include HEIC support (there was a separate github repo for windows builds with a manual).

**Reply by jcupitt** [reply]
Just check that you don't include imagemagick in the list of projects to build. It can have extremely high memory use for some formats, and has a history of security issues.

You can check with:

```
$ vips --vips-config
enable debug: true
enable deprecated: true
enable modules: true
enable cplusplus: true
enable RAD load/save: true
enable Analyze7 load/save: true
enable PPM load/save: true
enable GIF load: true
use fftw for FFTs: true
accelerate loops with ORC: true
ICC profile support with lcms: true
zlib: true
text rendering with pangocairo: true
font file support with fontconfig: true
EXIF metadata support with libexif: true
JPEG load/save with libjpeg: true
JXL load/save with libjxl: true (dynamic module: true)
JPEG2000 load/save with OpenJPEG: true
PNG load/save with libspng: false
PNG load/save with libpng: true
selected quantisation package: imagequant
TIFF load/save with libtiff: true
image pyramid save with libgsf: true
HEIC/AVIF load/save with libheif: true (dynamic module: true)
WebP load/save with libwebp: true
PDF load with PDFium: false
PDF load with poppler-glib: true (dynamic module: true)
SVG load with librsvg: true
EXR load with OpenEXR: true
OpenSlide load: true (dynamic module: true)
Matlab load with libmatio: true
NIfTI load/save with niftiio: true
FITS load/save with cfitsio: true
GIF save with cgif: true
selected Magick package: MagickCore (dynamic module: true)
Magick API version: magick6
Magick load: true
Magick save: true
```

I have it enabled on this libvips, so I probably wouldn't use it on potentially hostile files.

**Reply by andy250** [reply]
Hi @jcupitt 
I am trying to build it again for Windows x64 using this repo https://github.com/libvips/build-win64. Can you please give me some instructions on how to build it without imagemagick / exclude it from the project list? The `build.sh` script accepts only `web` or `all` (the latter includes imagemagick), and there are a few flags supported, but nothing related to imagemagick. I started wondering if maybe I should just go with `web` (instead of `all`) and `--with-hevc`.

**Reply by kleisauke** [reply]
ImageMagick is build as dynamic module within the `-all` variant, so you can just remove `bin/vips-modules-x.y/vips-magick.dll` if you want to disable load/save via libMagick. Alternatively, libvips >= 8.13 lets you disable any unfuzzed loaders from running via the `VIPS_BLOCK_UNTRUSTED` env var, regardless of which loaders are compiled into the libvips binary.

The `-web` variant is indeed recommended if you process untrusted images.

**Reply by andy250** [reply]
Oh, that's great, thank you @kleisauke.

**Reply by andy250** [reply]
@kleisauke I've built 8.14.2 now. After removing `bin\vips-modules\vips-magick.dll` and then running `vips --vips-config` I am still getting:
```
selected Magick package: MagickCore (dynamic module: true)
Magick API version: magick6
Magick load: true
Magick save: true
```
Is this normal? Any other way to make sure imagemagick is not used?

**Reply by jcupitt** [reply]
Yes, the `--vips-config` flag shows how libvips was configured at compile-time, not which operations are available at run-time.

Try `vips magickload` and see what happens. I see:

```
$ vips magickload
load file with ImageMagick
usage:
   magickload filename out [--option-name option-value ...]
where:
   filename     - Filename to load from, input gchararray
   out          - Output image, output VipsImage
optional arguments:
   density      - Canvas resolution for rendering vector formats like SVG, input gchararray
... 
operation flags: nocache untrusted 
$ rm ~/vips/lib/vips-modules-8.15/vips-magick.so 
$ vips magickload
vips: unknown action "magickload"
```

**Comment by andy250** [comment]
Hi @jcupitt thanks for the lightning-fast answer. It's rather random and not easy to track. Even if I find some offending files I won't be able to share them, but I'll surely look into `vipsheader` and try to find a pattern. Thanks for the suggestion!

**Reply by jcupitt** [reply]
Personally I think I would use a language binding like netvips or pyvips. You get a lot more flexibility in the kinds of preprocessing you can do (the command-line interface to libvips is deliberately weak). You could get some basic image stats and log them before starting a thumbnail, perhaps it would show something.

**Reply by andy250** [reply]
I started off with netvips, but I ran into some trouble and had to fall back to command line. It's a pity I don't remember exactly what the problem was with netvips.

**Comment by jcupitt** [comment]
... I made a more extreme example of a memory bomb:

```
$ vips black x.tif[compression=ccittfax4,bitdepth=1] 9999999 10000
$ ls -l x.tif
-rw-rw-r-- 1 john john 2330 May  8 14:06 x.tif
$ /usr/bin/time -f %M:%e vipsthumbnail x.tif 
25763840:518.84
```

So a 2kb image file that needs 25gb of memory and 10 minutes of CPU to thumbnail.

You can check for interlaced PNG and JPG with the `interlaced` metadata item:

```
$ vipsheader -f interlaced interlaced.jpg 
1
$ vipsheader -f interlaced k2.jpg 
vipsheader: vips_image_get: field "interlaced" not found
```

Interlaced images need a *LOT* more cpu and memory to process, so you might want to put them on a separate path. For example:

```
$ vipsheader st-francis.jpg
st-francis.jpg: 30000x26319 uchar, 3 bands, srgb, jpegload
$ /usr/bin/time -f %M:%e vipsthumbnail st-francis.jpg
67072:2.61
$ vips copy st-francis.jpg st-francis-interlaced.jpg[interlace]
$ /usr/bin/time -f %M:%e vipsthumbnail st-francis-interlaced.jpg 
2380252:7.07
```

Up from 2.6s to 7s of CPU, and up from 70MB to 2.4GB.

**Reply by andy250** [reply]
Thanks for this, much appreciated. So out of curiosity - is this considered normal behavior and there are no memory optimizations available here? By dumping some parts to disk maybe. Or is this coming from external libs you are out of control?

**Reply by jcupitt** [reply]
Unfortunately, yes, this is inevitable. There are some really terrible image formats out there (I'm looking at you, GIF), and they do simply need a lot of memory for some files. This is why imagemagick has `policy.xml`, the config file you can use to set things like image dimension limits.

libvips takes a different (and IMO better, of course) approach -- it makes `vipsheader` always quick, safe and low memory, and applications which process potentially hostile files should probably use that to do some amount of screening of content before processing. 

A check like `width * height * bands < 10000 * 10000 * 3` would be enough (if your users are OK with that).

**Reply by andy250** [reply]
Hey @jcupitt I am trying out the "vipsheader" approach. After applying it I detected high memory situation on one of the servers - it turned out it is caused both by vips.exe and vipsheader.exe. I am doing really lots of file processing, but the memory usage for vipsheader is unexpected, since you mentioned that it is always quick and low on memory. Also vipsheader is not completing quickly (CPU time reported by task manager was over 2 mins, and memory usage finally grew to 4GB per process, then I killed it). Here is win task manager screen:
![image](https://github.com/libvips/libvips/assets/10115397/97821bb8-8e93-4288-a61c-1a4ba134619f)

Any hints on next steps in my investigation? I will try to isolate problematic files, and try to come up with some conclusions. Probably I won't be able to share the files though.

**Reply by jcupitt** [reply]
Ouch!

There are some formats (or subtypes) which just can't be processed efficiently, perhaps you've hit one of them? Though `vipsheader` should always be fast, I'm puzzled you're having trouble with that.

Let me know what file type it's having difficulty with, and maybe the output of `vipsheader -a`. You could also try running something like `exiftool` or `tiffinfo` on it and see what output you get.

**Reply by jcupitt** [reply]
(so eg. interlaced PNG and JPG will always need a lot of memory to thumbnail)

**Reply by andy250** [reply]
Got it, thanks, I'll keep on digging and report back.

**Reply by jcupitt** [reply]
... you can get very high memory use if have imagemagick enabled and try to thumbnail a video file. But I think you've fixed this, so I'm puzzled.

For example:

```
$ /usr/bin/time -f %M:%e vipsheader 1986.avi 
vipsheader: magick2vips: libMagick error: cache resources exhausted `/tmp/magick-cUd-4pt3vPRAW68T5q0KhgINyLg2OIeF.pam' @ error/cache.c/OpenPixelCache/4095
magick2vips: unable to read file "1986.avi"
Command exited with non-zero status 1
37879552:59.15
$ 
```

So imagemagick tried to decode the entire video into memory (38gb!) before beginning to thumbnail it haha.

**Reply by andy250** [reply]
Yeah, after some more investigation it turns out I was indeed running vipsheader on some videos (in addition to images). So probably false alarm here :) Fix is on the way.

**Reply by jcupitt** [reply]
Hmm though that makes it seem like you have imagemagick enabled, since it's the only thing that will attempt to decode video. Have you tried running `vips magickload` to verify that it's not there?

**Reply by andy250** [reply]
ImageMagick is there (bad, I know) :) My current production build is 8.14.1 with imagemagick included. I have now ready build 8.14.2 locally and I can confirm vips magickload fails (`vips: unknown action "magickload"`) locally after removing the dlls mentioned by kleisauke. Thanks for the update from yesterday, only now that I saw it (https://github.com/libvips/libvips/discussions/3484#discussioncomment-6060284).

**Reply by jcupitt** [reply]
With 8.14 you can also set the env var `VIPS_BLOCK_UNTRUSTED` and libvips will block all untrusted loaders, so IM plus some others. It might be easier.

**Reply by andy250** [reply]
I tried this variable now. I was puzzled at first as this variable doesn't affect `vips magickload` but it correctly blocks `vipsheader video.mp4` - result is `vipsheader.exe: magickload: operation is blocked`. Thanks.
